subprojects/functional/src/main/java/org/gradle/internal/Either.java

```diff
@@ -34,9 +37,54 @@
         return new Right<>(value);
     }
 
-    public abstract <U> U fold(Function<L, U> l, Function<R, U> r);
+    /**
+     * Is this a left?
+     */
+    public abstract boolean isLeft();
 
-    public abstract <U> Either<L, U> map(Function<R, U> r);
+    /**
+     * Is this a right?
+     */
+    public abstract boolean isRight();
+
+    /**
+     * Take the value if this is a left.
+     */
+    public abstract Optional<L> ifLeft();
+
+    /**
+     * Take the value if this is a right.
+     */
+    public abstract Optional<R> ifRight();
+
+    /**
+     * Map either case.
+     */
+    public abstract <U, V> Either<U, V> map(Function<? super L, ? extends U> l, Function<? super R, ? extends V> r);
```

Comment:
I'd have the original `map` renamed to `mapRight` and introduce a separate `mapLeft` as this version of `map` is already covered by `fold<Either<U, V>>(..., ...)`.

Reply:
Done.